### PR TITLE
Remove hard-coded standard C++ library selection and add more releases in llvm package

### PR DESCRIPTION
**Pull request by ye-luo**
I'd like to add a new variant to allow using libstdc++ as default by the clang compiler even when libcxx is built.

### CODE REVIEW

**Comment by ye-luo** [comment]
@naromero77 ~what is the formatting rule/tool to fix the formatting CI failure?~ fixed

**Comment by naromero77** [comment]
@trws Can you take a look at this one?

**Comment by trws** [comment]
This is something I could see being useful, but I'd prefer to have the invariants expressed in it.  The main one is that you can't have the stdlib option set to libc++ if the variant to build libc++ isn't specified, as of now it would try to build it and probably fail in cmake or, worse, find one from some other build.

To your point @naromero77 I agree we're reaching a pretty heavy saturation on variants.  It might be worth taking a pass over them and seeing if they're all still relevant.  This one I'm in favor of if only because the behavior this offers is something we actually do on our deployed clang builds.

**Comment by ye-luo** [comment]
> The main one is that you can't have the stdlib option set to libc++ if the variant to build libc++ isn't specified, as of now it would try to build it and probably fail in cmake or, worse, find one from some other build.

I think the logic I put up is exactly what you expected, if +libcxx is not requested, clang_cxx_stdlib is completely ignored as before. Maybe I should make it more clear in the documentation.

**Comment by ye-luo** [comment]
> This one I'm in favor of if only because the behavior this offers is something we actually do on our deployed clang builds.

In almost all the desktop or HPC systems I used in the past, libstdc++ is the default since it is the default configuration choice when clang is built.

**Comment by trws** [comment]
> I think the logic I put up is exactly what you expected, if +libcxx is not requested, clang_cxx_stdlib is completely ignored as before.

This is not quite right.  As of now, it's an *error* not ignored.  I consider it to be an error to build a package with a variant saying it's using libcxx if libcxx is unavailable.

**Comment by ye-luo** [comment]
> > I think the logic I put up is exactly what you expected, if +libcxx is not requested, clang_cxx_stdlib is completely ignored as before.
> 
> This is not quite right. As of now, it's an _error_ not ignored.

Probably I don't understand spack enough and got more and more confused... Did you mean
`~libcxx clang_cxx_stdlib=not_default` should result in an error instead of getting clang_cxx_stdlib ignored?

> I consider it to be an error to build a package with a variant saying it's using libcxx if libcxx is unavailable.

could you be more specific, I don't quite understand what you mean here.

**Comment by ye-luo** [comment]
> as of now it would try to build it and probably fail in cmake or, worse, find one from some other build.

I don't get why it would try to build it?

**Comment by trws** [comment]
You’re right, it wouldn’t actually set anything, which is 
differently bad but still bad.

**Comment by trws** [comment]
That’s close to what I mean.  If the user sets 
clang_cxx_stdlib=libstdc++ that would be fine, but if they ask for a 
standard library that is not available that should not pass through as a 
valid variant.  In principle that also means it shouldn’t allow 
libstdc++ to be selected when only libc++ is available (as on Mac by 
default) but that’s less common and harder to test for, so just making 
sure libc++ actually exists when requested is probably enough for now.

**Comment by ye-luo** [comment]
The current behavior is
```
+libcxx
+libcxx clang_cxx_stdlib=default
+libcxx clang_cxx_stdlib=libc++
# builds libcxx and set default C++ standard library to libc++
+libcxx clang_cxx_stdlib=libstdc++
# builds libcxx and set default C++ standard library to libstdc++
~libcxx
~libcxx clang_cxx_stdlib=whatever_value
# libcxx is not built and clang default C++ standard library is set to libstdc++
```

I don't use mac so I cannot say much.
My understanding is llvm package [doesn't support a gcc tool chain from OS ](https://github.com/spack/spack/issues/16689)
So it is very rare that c++ is not enabled when gcc is built. So I assume libstdc++ is always available.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -418,8 +427,12 @@ def cmake_args(self):
         if "+libcxx" in spec:
```
Comment (ye-luo) [inline var/spack/repos/builtin/packages/llvm/package.py:445]: @trws are you saying this protection is not enough for clang_cxx_stdlib?

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 2]
```diff
@@ -86,6 +87,14 @@ class Llvm(CMakePackage, CudaPackage):
         default=True,
         description="Build the LLVM C++ standard library",
     )
+    variant(
+        "clang_cxx_stdlib",
+        default="default",
+        description="C++ standard library used by clang++. "
+        "Default is libc++ when +libcxx or libstdc++ otherwise. "
+        "The selection is effective only if +libcxx is specified",
+        values=("default", "libstdc++", "libc++"),
```
Comment (ye-luo) [inline var/spack/repos/builtin/packages/llvm/package.py]: Should I assume spack check input values against reference values at the first place?

**Comment by naromero77** [comment]
I just want to double-check that I understand the allowable configurations:

- Use libstdc++ (nothing to build).
   - Conflict for darwin is needed ??? Not familiar with Mac either
- Build libc++ and set libstdc++ as default
- Build libc++ and set libc++ as default

If these are the only allowable options, perhaps we should create a single multi-valued variant? It seems like having the two variants leads to an over complete set of options.

**Comment by trws** [comment]
We probably could.  At the build system level there are two options:

* Build libcxx? Yes/no
* Default stdlib? The one the compiler is built with, or new libcxx

On Mac, if you’re building llvm with the system compiler there is no 
libstdc++ available. If you’re building with a spack or home-brew 
built gcc, or a clang built with same, though then there is.  That’s 
why I think maybe we leave that one alone, it’s not a trivial check 
necessarily.

**Comment by naromero77** [comment]
OK, if there are two CMake options, then I would be OK with the two variants.

A few more questions:
- Does the clang_cxx_stdlib multi-valued variant really need a `default` option? There seems to be only two options `libc++` or `libstdc++` and some hard conflicts with the `+libcxx` variant. And if there are only two options, then it could just be a true/false-style variant.
- The naming here is a bit confusing `+libcxx` is the subproject that enables the creation of `libc++`, correct?

**Comment by trws** [comment]
It could be true/false in principle, but I'm not sure there are necessarily only two values (there probably are for now, but clang also supports Microsofts stl, which may at some point come up or some other).  I'm not sure I'd have a "default" as such as much as set the default to one of the libraries based on other variant selections personally, and then just have libstdc++ and libc++ as the variant values.

Yes, libcxx is the subproject that builds libcxx.  If that is built, the new clang defaults (in some cases anyway) to using that new libcxx as its c++ standard library rather than the one it was built with.

**Comment by naromero77** [comment]
So there would only be two values for the mult-valued variant. I *think* the changes look something like this:
```
  variant(
        "clang_cxx_stdlib",
        default="libc++",
        description="Default C++ standard library used by clang++.",
        values=("libstdc++", "libc++"),
```
and we default to `libc++` since we also default to `libcxx=True`

Plus we add a conflict:
`conflicts("clang_cxx_stdlib=libc++", when="~libcxx")`

**Comment by ye-luo** [comment]
when `+libcxx`, clang_cxx_stdlib needs default `libc++`
when `~libcxx`, clang_cxx_stdlib needs default `libstdc++`
I need a conditional default value. That is why I added this default option because I'm not aware of spack having such a feature.
Having a conflict is not user friendly.

**Comment by naromero77** [comment]
@adamjstewart @alalazo I still don't think its possible to have a conditional default value?

@ye-luo Yes, a conditional default value is better than a conflict. 

But this is also undesirable behaviour IMO, the variants are normally uniquely defined.

```
+libcxx clang_cxx_stdlib=default
+libcxx clang_cxx_stdlib=libc++
# builds libcxx and set default C++ standard library to libc++
+libcxx clang_cxx_stdlib=libstdc++
# builds libcxx and set default C++ standard library to libstdc++
~libcxx
~libcxx clang_cxx_stdlib=whatever_value
# libcxx is not built and clang default C++ standard library is set to libstdc++
```

You end up having to do:
`spack install llvm~libcxx clang_cxx_stdlib=libstdc++`

Not user friendly, but a limitation that we have to live with in Spack for now.

**Comment by naromero77** [comment]
Sometimes to make it a bit more user friendly, I will do
```
conflicts("clang_cxx_stdlib=libc++", 
               when="~libcxx",
               msg = "+libcxx is needed to use libc++ as the C++ standard library or set clang_cxx_stdlib=libstdc++")
```

**Comment by adamjstewart** [comment]
No conditional default variants, aside from OS-specific things.

**Comment by trws** [comment]
So, where are we stuck here? I'm leaning toward having the default be to have the default based on the OS, Mac gets libc++ and everything else gets libstdc++ by default.  Then include the conflict so a user can tell when they've asked for something that will not work.  @adamjstewart will that work with what spack can provide right now?

**Comment by adamjstewart** [comment]
If the only options are libc++ (macOS) and libstdc++ (everything else) then why do we need a variant at all?

**Comment by trws** [comment]
Those are sane defaults, since the base system matches those, but you can use either on either if you build them.

**Comment by adamjstewart** [comment]
I'm fine with using `default='libc++' if sys.platform == 'darwin' else 'libstdc++'`, but you're the maintainers so it's really up to you.

**Comment by ye-luo** [comment]
> So, where are we stuck here? I'm leaning toward having the default be to have the default based on the OS, Mac gets libc++ and everything else gets libstdc++ by default. Then include the conflict so a user can tell when they've asked for something that will not work. @adamjstewart will that work with what spack can provide right now?

So everyone is OK with a change of behaviour? Before libcxx is default when +libcxx is on. I'm OK with a change of behaviour.

**Comment by trws** [comment]
It is a behavior change, but it's the only way I can think of to avoid having `llvm clang_cxx_stdlib=default` and `llvm clang_cxx_stdlib=libstdc++` be different when they build the same thing and get something that works by default in the majority of cases.

**Comment by ye-luo** [comment]
@trws @naromero77 removed the new variant and removed forced use of libcxx in clang when libcxx is enabled. The default C++ standard library of Clang will be decided by Clang instead of spack. LLVM developers told me the default on Windows is MS VC C++ standard library. So it is already OS dependent.

**Review by naromero77 — APPROVED** [review]

**Comment by naromero77** [comment]
@trws I did not realize we where forcing a particular C++ library. I think Ye's change make sense and no new variant is introduced.

**Comment by ye-luo** [comment]
@trws Do you have merge rights or need the admin to do the merge?

**Comment by trws** [comment]
I do not have merge rights, but I agree this works, thanks!  @adamjstewart could we get a merge here?

**Review by adamjstewart — APPROVED** [review]

**Comment by trws** [comment]
Thanks!

**Comment by cdfh** [comment]
This change has caused me much stress. This is a silent change in behaviour, which I believe a project like spack should be strongly trying to avoid (frequent changes in behaviour, especially silent, anger users).

In addition, I would strongly argue that it is a bug to default against libc++ when the +libcxx variant is present: giving +libcxx strongly implies that the user wants libc++ _support_. But because you can't mix and match standard library implementations, even across libraries, it makes no sense to have libc++ support unless it's being used _everywhere_. 

Consider my use case, which doesn't feel like it's unusual/odd:

1. I compile `llvm+libcxx+clang`
2. I discover/configure it with `spack compiler find`
3. I install libraries with `%clang`
4. I compile my own project, managed by spack (formerly using `spack setup`), with `%clang`

I want my own project to use libc++. But, it makes absolutely no sense for me to compile my project with `-stdlib=libc++` because if I do, then all hell breaks loose due to mixing and matching standard libraries (any library my project depends upon will pull in `libstdc++`).

When `libc++` was the default, the user could still override it in `compilers.yaml` or via `-stdlib`, so it's not like the original behaviour was overly restricting. But if you're not wanting to use it, why build llvm with `+libcxx` at all?

But regardless, the main thrust of my argument is that this change silently changed behaviour and that this will cost users a great deal of time and energy as their ecosystems break as a result. I would thus propose that this change is harmful and should be reverted.

**Comment by ye-luo** [comment]
If you build LLVM with clang and libcxx enabled via CMake directly, clang still default to use libstdc++, on Linux unless CMake option `-DCLANG_DEFAULT_CXX_STDLIB=libc++` is explicitly set to request libc++. I believe LLVM has solid reason not to automatically default on using libc++. So we just followed LLVM choice. Similarly, you may install clang and libc++ from package manager and clang won't automatically pick libc++.

Enable libc++ in LLVM and using libc++ by default in clang are separate logic. `+libcxx` means libc++ is available to use but it is users' choice to opt-in when compiling codes. It is also bad to compile two versions of llvm+clang with the only difference of picking libstdc++ and libc++ as default in order to serve users of these two needs.

I think to address your issue, we can have a variant which exactly does setting `-DCLANG_DEFAULT_CXX_STDLIB=libc++` on demand. In such away, we have serve all the three use cases

1. clang select standard c++library based on llvm default, libc++ is compiled and available for use.
2. clang select standard c++library based on llvm default, no libc++ being compiled. No need of another clang build, it is covered by 1 already.
3. clang select libc++ as c++ library, libc++ is compiled.

**Comment by cdfh** [comment]
> So we just followed LLVM choice. 

That may be so, but it changed spack's behaviour and regressed a previous bug report (##5942, #5943). Spack has a large number of users, many using it in production environments. Changing its behaviour should be considered a major event not to be undertaken lightly. In this PR, there was another option: introduce a multi-valued `stdlib` variant, defaulting to the existing behaviour (which was actually your original proposal, which would be fine). Had this been done, no existing infrastructure's would have been broken. 

> I think to address your issue, we can have a variant which exactly does setting -DCLANG_DEFAULT_CXX_STDLIB=libc++ on demand. In such away, we have serve all the three use cases

The issue here isn't how to make it work (I've already addressed that locally). The issue is that the behaviour changed and that that resulted in breakages in at least one (and probably many more)  spack deployment. (In my case, Qt suddenly failed to compile for cryptic reasons due to mismatches in the standard library; this change introduced build errors in qt%clang.) For me, the price has already been paid on that issue and there's no fixing it. But for others, it hasn't been paid yet and the solution is to revert back to the existing behaviour and introduce this PR in a way that doesn't change previous defaults.

**Comment by haampie** [comment]
> But for others, it hasn't been paid yet 

This was merged a year and a half ago and part of two 0.x releases, with no complaints. The previous default was arguably highly unexpected, I don't know any distro that defaults to `libc++`.

**Comment by cdfh** [comment]
Well, no, there's now been one complaint. And it presumably also affects the original poster of #5943, where the default behaviour was originally introduced under the oversight of spack's original author. 

I don't update spack very often. Indeed, it has evidently been over a year and a half since I updated. Instead, I update individual packages in the package database when new versions are needed. The reason I don't update frequently is because each time I update spack, things break. Like this. I'm presumably not the only user who takes this workflow, so the fact that only two people are known to be affected over a 1.5 year period shouldn't be taken as evidence that problems haven't been more widespread.

**Comment by ye-luo** [comment]
Imposing rule like "Not making any breaking changes" just doesn't work in most software development although no breakage is always desired. I don't feel I'm in any position to discuss policy, let's move a bit to consider what we can do now. As indicated in the whole discussion history of this PR, the old behavior is not desired. I think adding a variant for setting libc++ as default can be a candidate to help users like @cdfh .

**Comment by cdfh** [comment]
> Imposing rule like "Not making any breaking changes" just doesn't work in most software development 

I can count the number of times that git, emacs, llvm, gcc, or cmake changed their behaviour in a way that affected my tooling on a single hand, so it is certainly possible. Very occasionally, cmake does, but when it does, it only does so after having first deprecated the original behaviour (giving verbal warnings at runtime) across multiple versions. 

I would agree with your premise that backwards compatibility isn't feasible for general applications, but spack is tooling rather than an application. People write infrastructures around it. When spack changes, those infrastructures break. So even if it can't adhere to a strict policy of backwards compatibility, it should strive towards it as much as is feasible. I would argue that in this case, it would have been very feasible to satisfy @ye-luo's original feature request without breaking existing infrastructures, just by adding a new variant, as originally proposed, instead of changing the meaning of an existing variant.

But if rolling back this change isn't welcomed, then I don't propose any further changes as it _is_ possible to work around this issue for users who want libc++ by modifying compilers.yaml. My complaint isn't really the functionality, but rather the change in behaviour.

**Comment by trws** [comment]
@cdfh, I understand your position, but taking this position as a complaint about a breaking change *a year and a half ago* is not reasonable.  If you would like to open an issue, or better pull request, to discuss having `+libcxx` use it by default for compiling c++ code with clang, please feel free to do so, but this is not the place for it.

> I want my own project to use libc++. But, it makes absolutely no sense for me to compile my project with `-stdlib=libc++` because if I do, then all hell breaks loose due to mixing and matching standard libraries (any library my project depends upon will pull in `libstdc++`).

That's true if you do it in your project, but not if you specify the flags to spack, which will build a new tree with those flags and generate a consistent environment for you.  

> In addition, I would strongly argue that it is a bug to default against libc++ when the +libcxx variant is present: giving +libcxx strongly implies that the user wants libc++ _support_. But because you can't mix and match standard library implementations, even across libraries, it makes no sense to have libc++ support unless it's being used _everywhere_.

Feel free to report this bug upstream on llvm, where this is what happens by default when libcxx is enabled.  We are passing that behavior on, not producing it.

> When `libc++` was the default, the user could still override it in `compilers.yaml` or via `-stdlib`, so it's not like the original behaviour was overly restricting.

Exactly like it is now.

> But if you're not wanting to use it, why build llvm with `+libcxx` at all?

Frequently I build with libcxx so that I can build a specific project with it, or statically link libcxx into a project that needs to have no visible stdlib dependency at runtime.  I also do this as part of multi-stage compiler bootstrapping where I want to build a compiler that will use libstdc++ but will then be used to build another that will be built with libc++.

> it presumably also affects the original poster of https://github.com/spack/spack/pull/5943, where the default behaviour was originally introduced under the oversight of spack's original author.

That bug was fixed, in a way, by the patch but only in that the resulting clang could run.  The underlying issue was that rpathing for compiler standard libraries were not being handled correctly by spack yet, so building clang without libcxx would have created the same error the same way.  He would not have had that issue today.  

> I would agree with your premise that backwards compatibility isn't feasible for general applications, but spack is tooling rather than an application. People write infrastructures around it. When spack changes, those infrastructures break. So even if it can't adhere to a strict policy of backwards compatibility, it should strive towards it as much as is feasible. I would argue that in this case, it would have been very feasible to satisfy @ye-luo's original feature request without breaking existing infrastructures, just by adding a new variant, as originally proposed, instead of changing the meaning of an existing variant.

That is a very fair point, and perhaps we should have done it that way.  But I am not going to accept making *another breaking change* to fix a complaint about a breaking change.  Doing so now would break far more people than leaving it because reuse support will cause recently built binaries to be reused more often and the issue will come up faster and more frequently.

**Comment by adamjstewart** [comment]
Just adding my two cents:

Spack is still in beta release. Until we have a 1.X release, users should expect backwards-incompatible changes from time to time. For major things like removing versions/packages/commands, we always [deprecate](https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions) them first and remove them after at least one full release cycle. We don't yet have any way to deprecate variants, but I would like to add that.

In general, Spack has so many packages that it's almost impossible to avoid these kinds of changes in every package. In order to make packages more stable, we added a "maintainer" system where users can register themselves as willing to help maintain and test a particular package recipe. For example, I'm personally maintaining all Python libraries in Spack, particularly geospatial and ML packages. When a user opens an issue or PR relating to a package I maintain, I'm automatically notified and asked to review.

The core maintainers of Spack (the people with merge privileges) can't possibly know everything about every package in Spack. When someone proposes a change like this in a PR, a developer like me has to trust the "maintainers" of that build recipe to make the appropriate decision. If a package has no official "maintainer", I just merge PRs willy nilly even if they involve a change in default behavior.

TL;DR: if you have packages like LLVM that you rely on and you would be willing to help maintain those build recipes, I would encourage you to add yourself to the list of "maintainers" for that `package.py` so that you can add your opinions on these kind of changes.

**Comment by cdfh** [comment]
@trws 

> But I am not going to accept making another breaking change to fix a complaint about a breaking change. 

That's fair enough, in which case please disregard my earlier proposal to do so.

However, I very much stand by that this _is_ the place to raise the issue that this change should not have been made at the time, and I believe that point needs to be made in order to raise awareness of the need for greater stability. If nobody complains, then the spack maintainer community presumes everything is fine (which was exactly @haampie's position from earlier: nobody complained, therefore no foul). 

I've made a considerable number of bugfixes to spack. But none of them have been shared upstream, which is tragic. The reason? My branch is often very far behind (as observed, over a year behind until this week). Once I track down and fix a spack bug (which is normally a very time-consuming process), I do not then have time to contribute upstream because doing so would require forward porting it to origin/HEAD, which after an already long bug hunt, I simply do not have time to do. If my branch was already at origin/HEAD (or close to it), then contributing back upstream would be easy. But previous efforts to keep my branch up to date resulted in way too frequent failures, and so now I do it very rarely (I have a snapshot that works, and only pull when I really, really need to).

So, I withdraw my proposals for reverting with respect to the current PR, but please take note: when behaviour changes unexpectedly, it breaks things for users and creates challenges that will prevent some users from contributing back to spack's development.

**Comment by adamjstewart** [comment]
> I've made a considerable number of bugfixes to spack. But none of them have been shared upstream, which is tragic. The reason? My branch is often very far behind (as observed, over a year behind until this week).

An approach I recommend is to have a "production copy" of Spack that sticks to stable releases and a "development copy" of Spack in your home directory that you keep up-to-date with develop. This worked for me for many years as a HPC sys admin at ANL.

**Comment by cdfh** [comment]
@adamjstewart 

> For major things like removing versions/packages/commands, we always deprecate them first and remove them after at least one full release cycle. 

That is absolutely perfect and I support that. If infrastructure breaks because a functionality vanished, it's immediately recognised and pretty easily worked around. For example, we heavily relied upon `spack setup`, and when one day it vanished, it was well documented what the alternative was and it took minutes to recognise the problem and recover.

However, this change, by contrast, didn't remove functionality, but changed existing functionality in a way that had very subtle repercussions. Instead of breaking infrastructure in an obvious "command does not exist" way, it took literally days and a lot of stress to hunt down what was happening (qt was mixing standard library versions in a way that required manual interpretation of nm dumps to track down). Obviously, nobody anticipated at the time of creating this PR that it would cause such stress, but the point I'm making is that the original request suggested a fix that wouldn't have changed existing behaviour (a new variant), but this direction was abandoned because there are already a lot of variants and it would be cleaner/better to change behaviour. While from a design perspective that may be so, but from a software maintenance perspective I plead, please consider for a moment that _any_ time functionality is changed, it _will_ affect some percentage of users in esoteric ways, so please, if an alternative exists which is almost equally equivalent (such as in this case). do not turn away from said alternative lightly.
